virtual_rainforest/models/soil/carbon.py

```diff
@@ -0,0 +1,201 @@
+"""The `soil.carbon` module.
+
+This module simulates the radiation soil carbon cycle for the Virtual Rainforest. At the
+moment only two pools are modelled, these are low molecular weight carbon (LMWC) and
+mineral associated organic matter (MAOM). More pools and their interactions will be
+added at a later date.
+"""
+
+import numpy as np
+from numpy.typing import NDArray
+
+from virtual_rainforest.core.logger import log_and_raise
+from virtual_rainforest.core.model import InitialisationError
+
+# from core.constants import CONSTANTS as C
+# but for meanwhile define all the constants needed here
+BINDING_WITH_PH = {
+    "slope": -0.186,  # unitless
+    "intercept": -0.216,  # unitless
+}
+"""From linear regression (Mayes et al. (2012))."""
+MAX_SORPTION_WITH_CLAY = {
+    "slope": 0.483,  # unitless
+    "intercept": 2.328,  # unitless
+}
+"""From linear regression (Mayes et al. (2012))."""
+MOISTURE_SCALAR = {
+    "coefficient": 30.0,  # unitless
+    "exponent": 9.0,  # unitless
+}
+"""Used in Abramoff et al. (2018), but can't trace it back to anything more concrete."""
+TEMP_SCALAR = np.array([15.4, 11.75, 29.7, 0.031, 30.0])
+"""Used in Abramoff et al. (2018), but can't trace it back to anything more concrete.
+
+Values:
+    t_1: (degrees C)
+    t_2: (unit unclear)
+    t_3: (unit unclear)
+    t_4: (unit unclear)
+    ref_temp: (degrees C)
+"""
+
+
+class SoilCarbonPools:
+    """Class containing the full set of soil carbon pools.
+
+    At the moment, only two pools are included. Functions exist for the transfer of
+    carbon between these pools, but not with either the yet to be implemented soil
+    carbon pools, other pools in the soil module, or other modules.
+    """
+
+    def __init__(self, maom: NDArray[np.float32], lmwc: NDArray[np.float32]) -> None:
+        """Initialise set of carbon pools."""
+
+        # Check that arrays are of equal size and shape
+        if maom.shape != lmwc.shape:
+            log_and_raise(
+                "Dimension mismatch for initial carbon pools!",
+                InitialisationError,
+            )
+
+        # Check that negative initial values are not given
+        if np.any(maom < 0) or np.any(lmwc < 0):
+            log_and_raise(
+                "Initial carbon pools contain at least one negative value!",
+                InitialisationError,
+            )
+
+        self.maom = maom
+        """Mineral associated organic matter pool"""
+        self.lmwc = lmwc
+        """Low molecular weight carbon pool"""
```

Comment:
These strings describing the attributes should go in the Class docstring (to document the attributes) and in the `__init__` docstring to document what those parameters mean.

Reply:
Shouldn't there usually only be one of those two things? It seems to be common practice to either document the class in the class docstring or in the class.__init__ docstring, but not to have both?

@alexdewar raised the question of where to put attribute docstrings [here](https://github.com/ImperialCollegeLondon/virtual_rainforest/pull/140#discussion_r1064568241)  - we just need to agree on a common practice.

Reply:
Although I note the Google pystyle still has a stub doc for `__init__`:
https://google.github.io/styleguide/pyguide.html#384-classes

Reply:
@alexdewar suggestion is correct and so is mine according to the [Google Style Docstring guide for python](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html). So, as you say, it is a question of agreeing on the common practice. I prefer my option, obviously :P

Reply:
Just check that the option you choose renders as you would like in Read The Docs!

Reply:
Have checked that these render okay in Read The Docs. Personally I prefer the style with docstrings after each attribute as I feel it makes it easier to notice when they are missing. What do others think @davidorme @vgro @TaranRallings @robewers01 ?

Reply:
Shall we discuss this in person next week. We can put together a few examples of different structures and see what we feel about the rendering and code style.

In the past, I have generally put `__init__` is documented in the class docstring. It then appears naturally in docs right where the class is introduced. I think I prefer documenting the instance attributes using docstrings in  `__init__` (or above `__init__` for class variables). It is clearer to see when one is missing and the list of attributes in the compiled docs is built automatically.

I'm not convinced about that model - having a class overview in the class docstring and then `__init__` specific content with `__init__` also works.

```python
class MyClass:
    """This is my class.

    There are many like it, but this one is mine. Lots of text.

    Args:
        myarg: A value for my class
    """"

    unknown: int = 0
    """Need to explicitly say this is a class attribute - it gets lumped in."""

    def __init__(self, myarg: int) -> None:
        """Create a new MyClass instance."""

        self.myarg: int = myarg
        """The myarg value used to create the instance."""
        self.twice_myarg: int = self.myarg * 2
        """Twice my arg"""
```

Reply:
Yeah probably a sensible thing to decide in person. I guess I can leave them as they are for now, and then if we decide to get rid of them I can remove them during the session next week?